api/src/main/java/net/kyori/adventure/text/width/PixelWidthSourceImpl.java

```diff
@@ -0,0 +1,368 @@
+   * english alphanumerics and most punctuation and handle {@link TextDecoration#BOLD} in the style.
+   * See {@link PixelWidthSourceImpl#DEFAULT_FONT_WIDTH} for an example.</p>
+   *
+   * @param function a function that can provide a {@link CharToIntFunction} given a context
+   * @since 4.5.0
+   */
+  PixelWidthSourceImpl(final @NonNull Function<CX, CharToIntFunction<Style>> function, final @NonNull Function<CX, Locale> localeFunction){
+    this.function = function;
+    this.localeFunction = localeFunction;
+  }
+
+  @Override
+  public int width(final @NonNull Component component, final @NonNull CX context) {
+    final Map<String, Style> map = this.serialize(component, context);
+    final AtomicInteger length = new AtomicInteger(0);
+
+    map.forEach((string, style) -> string.chars().forEach(c -> length.getAndAdd(this.width((char) c, style, context))));
+
+    return length.get();
+  }
+
+  @Override
+  public int width(final @NonNull String string, final @NonNull Style style, final @NonNull CX context) {
+    int length = 0;
+    for(final char c : string.toCharArray()) length += this.width(c, style, context);
+    return length;
+  }
+
+  @Override
+  public int width(final char c, final @NonNull Style style, final @NonNull CX context) {
+    return this.function.apply(context).applyAsInt(c, style);
+  }
+
+  @SuppressWarnings(value="unchecked")
+  @Override
+  public <CO extends Component> void addResolver(final @NonNull Class<CO> resolveFor, final @NonNull ComponentResolver<CO, CX> resolver) {
+
+    if(resolveFor.isAssignableFrom(TextComponent.class)) {
+      throw new UnsupportedOperationException("Can not add custom resolver for TextComponents");
+    } else if(resolveFor.isAssignableFrom(TranslatableComponent.class)) {
+      this.translatableResolvers.add((ComponentResolver<TranslatableComponent, CX>) resolver);
+      return;
+    } else if(resolveFor.isAssignableFrom(KeybindComponent.class)) {
+      this.keybindResolvers.add((ComponentResolver<KeybindComponent, CX>) resolver);
+      return;
+    } else if(resolveFor.isAssignableFrom(ScoreComponent.class)) {
+      this.scoreResolvers.add((ComponentResolver<ScoreComponent, CX>) resolver);
+      return;
+    } else if(resolveFor.isAssignableFrom(SelectorComponent.class)) {
+      this.selectorResolvers.add((ComponentResolver<SelectorComponent, CX>) resolver);
+      return;
+    } else if(resolveFor.isAssignableFrom(NBTComponent.class)) {
+      if(resolveFor.isAssignableFrom(BlockNBTComponent.class)) {
+        this.blockNBTResolvers.add((ComponentResolver<BlockNBTComponent, CX>) resolver);
+        return;
+      } else if(resolveFor.isAssignableFrom(EntityNBTComponent.class)) {
+        this.entityNBTResolvers.add((ComponentResolver<EntityNBTComponent, CX>) resolver);
+        return;
+      } else if(resolveFor.isAssignableFrom(StorageNBTComponent.class)) {
+        this.storageNBTResolvers.add((ComponentResolver<StorageNBTComponent, CX>) resolver);
```

Comment:
The more I look at this, the more I think `width` should just be method of `TextComponent`. I think the current approach (in totality) is overly complex. It's not practical to calculate the `width` of non-text.

Reply:
hm that would play well into my hope for "two systems" -- some system for component 'flattening' is still important to have, then this system for width calculations.

Though -- since text components can have non-text children, I don't think such a method could be safely put on TextComponent

Reply:
Width is only calculated on text, but other components may need rendering before text representative of their content can be displayed. The resolvers added here does not calculate anything, they just open the possibility of resolving components so the text they represent can be used for a precise calculation.